### ISSUE TITLE
"Post Types" vs "Content Types"

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/hooks-api#29]
We need to standardize on calling post types either "Post Types" or "Content Types". Currently we use the term "Content" to refer to posts generically in some places, but we still use the title "Post Type" for the Post Type Entity.

The reason that there is such a dilemma is that the term "post" becomes ambiguous when it can be used to refer to the Post post type specifically as well as to all post types generally.

Custom post types are something that users are becoming more familiar with, but which new users may not understand. They aren't different types of Posts (in the sense of the Post content type), but are different types of content. For this reason some people refer to them as custom content types, and I think I prefer that. At least it seems to me that the term "custom content types" is something the user can more quickly and easily grasp.

I cannot find a reference right now, but I remember reading a comment (maybe from Nacin?) that custom post types should never have been exposed to the user under that name. But at this point is it a loosing battle?

**Comment by JDGrimes** [comment]
I think it is really pretty simple: anyone, no matter how well versed in WordPress's code or how new to it as a user, can probably grasp the term "Content Type". But only the more experienced will be able to grasp the term "Post Type". So its a no-brainer, right?

If only life were simple. The term "content" is even more ambiguous than the term "post", because it can apply to all kinds of things that don't get saved in the `posts` table. It can also lead to using the same term for the post ("Content") and the post's, er, well, content.

Also, the term "publish" may not exactly fit across content types, though I guess it should for the vast majority. 

## Multiple Events

I have, from time to time, fleetingly considered splitting out the generic Post Publish (or Content Publish) event for a different event for each content type. The upside to doing that is that it would let us be more fine-grained, custom-tailoring each event title to match the content type it was for. However, it also means more work for both the developer and the user. We could conceivably automatically register an event for each post type that is registered. But then the user can no longer award points across content types with a single reaction. Though I'm not sure whether that is even a useful feature.

There would also be other things to consider. Like what happens when a post type is deregistered. Right now, were that to happen, the reactions would remain visible, but would cease to be triggered. However, with the events split, when a post type was removed the event would cease to be registered, making the reactions still there, but dormant and hidden. It could be unexpected for them to suddenly come back if the post type was added later.

Unfortunately though, that wouldn't even solve the problem on its own. Because we'd need to do more than just register separate events, we'd also need to register separate Hook Args for each of them (they could continue to use the same Entities and Actions though). Or maybe it would be possible to use just a single Hook Arg for all of them but supply the Arg title dynamically—I'd have to investigate. Anyway, the point is that the Entity would still be called "Post" or "Content" in the UI for all of those events unless a custom Arg was used.

But that reminds me, the Entity will also be called "Post" or "Content" across all of the non-post_publish events anyway—so we still have to decide what to call, because we can't possibly split every event (and it actually wouldn't help in many cases).

### Conclusion

Splitting the Event isn't necessarily entirely off the table, but it is a tangent to this ticket, and should have a separate issue opened for it if we decide we want to explore it further. It is something that we can decide to do sometime later, we don't need to do it for the initial version.

**Comment by JDGrimes** [comment]
A [search on translate.wp.org](https://translate.wordpress.org/projects/wp/dev/en-ca/default?filters[term]=post+type&filters[user_login]=&filters[status]=current_or_waiting_or_fuzzy_or_untranslated&filter=Filter&sort[by]=priority&sort[how]=desc) reveals that the term "post type" isn't really exposed in the UI (most of the uses are in the XML-RPC class).

**Comment by JDGrimes** [comment]
The potential for confusion between "content types" as in post types and just "content" in general isn't a problem. However, the ambiguity between the "Content" as in Post and "Content" as in the post's content is an issue. I don't think that it is possible to resolve it (we can't change what we call the post's content, after all). Are there other options here that we are missing?

I just happened to think, using the term post has another drawback that I hadn't considered before. Because if we call it the "Post Publish" hook, users will be surprised when points are suddenly getting awarded for other post types too. Previously this wasn't as bad a problem, because the post type selector was displayed there, indicating that there were multiple types of posts that could be awarded. But now we aren't displaying that by default, the user has to discover it themselves by playing with the conditions, selecting the Type, and selecting the Equals condition. Only once they've added that condition will they see the dropdown to select different points types.

**Comment by JDGrimes** [comment]
Of course, we could explain that the Publish Post event includes other post types too in the description of the event in the UI. But again, that would only mitigate the confusion for this event, it wouldn't solve the issue of what to call the post types in other events.

---

I'm still trying to see if I can think of other words which we could use. Like "item" for example. But that is too broad since it could apply to too many things.

---

Another problem I thought of with the term "Content" is that it doesn't really have a good plural, should we ever need to use the plural form. "Contents" doesn't really convey "multiple items", but "the contents of an item". Its kind of strange I guess, but "content" can actually be used as a plural (more ambiguity).

---

Another possible word would be "article", but that wouldn't work for us either.

---

I'm sort of leaning toward "post" now, but I can't say I'm satisfied with the potential confusion for users. It just isn't ideal.

**Comment by JDGrimes** [comment]
## Splitting the Entity

One option is to register a different entity type for each type of post. Then we'd register the publish event for each of the post types.

### Benefits

This would also give us more fine-grained control over what events are registered for which post types. And I think that is significant, since some events will only apply to specific post types, not all of them. (We can already sort of do this by supplying the custom post type's name as the title of the event argument, however, the "Type" selector would still be displayed with all of the post types, which would be strange.)

### Problems

#### Multi-post type Reactions

One drawback of course would be that we can't use the same reaction for multiple post types. But as I said above, I'm not sure if that is really a useful feature anyway. (Right now it isn't as useful as it could be, because it is an all or single thing; you can't select a specific 2 or 3 or n post types to award while excluding the others.) There would be a work-around, however, in that custom events could be registered with custom entity types that cover multiple entities.

#### Duplication

Another big drawback is that we'd have to duplicate registration of all of the Entity children (atts, etc.) for each type of post. This is actually a benefit, having more fine-grained control, because we can then register children just for a specific type of post. (This would be useful for metadata/custom fields. We could supply a generic custom fields relationship, but users would have to do the dirty-work of entering unfriendly meta keys.) Also, the registering of the basic children could be easily placed within a loop which would register for all the types of posts without duplicating code. Another possibility would be to utilize some sort of alias system, by which the registration would only have to occur once.

#### Backward Compatibility

Another small consideration is back-compat. When we import the old points hooks, any that are awarding for all types of posts will be split into new reactions for each post type publish event. This will result in a slight change of behavior, in that new post types that are added later won't automatically get awarded points. Of course, we could even go so far as to keep track of this and add reactions for new post types in that situation, but I think that would be a worse user experience than just breaking back-compat here.

**Comment by JDGrimes** [comment]
#### Entity multiplication

Another issue that we have to consider with splitting the entities is that there will be more of them. Only about 2 more on a vanilla install, but perhaps many more on installs with many custom post types. This means that the events will also multiply accordingly.

**Comment by JDGrimes** [comment]
> ### Benefits
> 
> This would also give us more fine-grained control over what events are registered for which post types. And I think that is significant, since some events will only apply to specific post types, not all of them. 

This is also true for entity children, which means we can register a taxonomy only as a child of those post types that it supports. A big improvement, IMO.

> (We can already sort of do this by supplying the custom post type's name as the title of the event argument, however, the "Type" selector would still be displayed with all of the post types, which would be strange.)

Just to clarify, if we register entities for each post type we won't have issues with the Type argument being displayed. We'd have no need to register the Type as a child of each post type.

**Comment by JDGrimes** [comment]
> ### Problems

One more, mentioned above:

#### Ghost Reactions

> [W]hat happens when a post type is deregistered[?] Right now, were that to happen, the reactions would remain visible, but would cease to be triggered. However, with the events split, when a post type was removed the event would cease to be registered, making the reactions still there, but dormant and hidden. It could be unexpected for them to suddenly come back if the post type was added later.

Actually, even as things stand, all is not rosy with the world. The reactions will be displayed, and if the user attempts to edit them they could easily cause the post type to be changed. That is, the post type will no longer be registered, so it will not be displayed in the dropdown. This could lead to the user unwittingly changing the setting.

So, either way we go, we should probably introduce some sort of "dormant" mode for reactions, where they can be viewed and deleted but not edited. These are edge cases though, and I think they can wait.

**Comment by JDGrimes** [comment]
#### Event Multiplication

I was just contemplating this some more, and I happened to think of another side effect of splitting the entities: the comment events would have to be split too. This has its benefits of course, because not all post types allow commenting anyway. But it just demonstrated to me how the splitting will have farther-reaching implications than one might have thought.

Basically, any event which is for posts or entities which are related to posts (comments/taxonomies/terms/ect.) will have to split as well. Or, OK, maybe it wouldn't _have_ to, but it would sort of undermine the points of doing this in the first place: to avoid referring to post types generically.

All in all, splitting the comment events really seems like a positive, because it allows us to provide better names for the events.

I guess really what this all implies is that we would be splitting the comment entities as well. Which makes sense: **We'd have to split any entity which relates to posts, because there will no longer be a single generic `post` entity that it can relate to.** Instead there will be `post`, `page`, `media`, etc.

**Comment by JDGrimes** [comment]
Closing in favor of #42.
